contracts/InterestRate/StableRateModel.sol

```diff
@@ -0,0 +1,129 @@
+    }
+
+    /**
+     * @notice Updates the parameters of the interest rate model (only callable by owner, i.e. Timelock)
+     * @param baseRatePerYear_ The approximate target base APR, as a mantissa (scaled by BASE)
+     * @param stableRatePremium_ The multiplierPerBlock after hitting a specified utilization point
+     * @param optimalStableLoanRatio_ Optimal stable loan rate percentage.
+     * @notice Emits NewStableInterestParams, after updating the parameters
+     * @notice Only governance
+     */
+    function updateStableRateModel(
+        uint256 baseRatePerYear_,
+        uint256 stableRatePremium_,
+        uint256 optimalStableLoanRatio_
+    ) external {
+        require(msg.sender == owner, "StableRateModel: only owner may call this function.");
+
+        updateStableRateModelInternal(baseRatePerYear_, stableRatePremium_, optimalStableLoanRatio_);
+    }
+
+    /**
+     * @notice Calculates the ratio of the stable borrows to total borrows
+     * @param stableBorrows The amount of stable borrows in the market
+     * @param totalBorrows The amount of total borrows in the market
+     * @return The stable loan rate as a mantissa between [0, BASE]
+     */
+    function stableLoanRatio(uint256 stableBorrows, uint256 totalBorrows) public pure returns (uint256) {
+        // Loan ratio is 0 when there are no stable borrows
+        if (totalBorrows == 0) {
+            return 0;
+        }
+
+        return (stableBorrows * BASE) / totalBorrows;
+    }
+
+    /**
+     * @notice Calculates the current borrow rate per block, with the error code expected by the market
+     * @param stableBorrows The amount of stable borrows in the market
+     * @param totalBorrows The amount of borrows in the market
+     * @param variableBorrowRate Current variable borrow rate per block of the protocol
+     * @return The borrow rate percentage per block as a mantissa (scaled by BASE)
+     */
+    function getBorrowRate(
+        uint256 stableBorrows,
+        uint256 totalBorrows,
+        uint256 variableBorrowRate
+    ) external view returns (uint256) {
+        uint256 loanRatio = stableLoanRatio(stableBorrows, totalBorrows);
+        uint256 excessLoanRatio = calculateLoanRatioDiff(loanRatio);
+
+        return (variableBorrowRate + baseRatePerBlock + ((stableRatePremium * excessLoanRatio) / BASE));
+    }
+
+    /**
+     * @notice Internal function to update the parameters of the interest rate model
+     * @param baseRatePerYear_ The approximate target base APR, as a mantissa (scaled by BASE)
+     * @param stableRatePremium_ The multiplierPerBlock after hitting a specified utilization point
+     * @param optimalStableLoanRatio_ Optimal stable loan rate percentage.
+     */
+    function updateStableRateModelInternal(
+        uint256 baseRatePerYear_,
+        uint256 stableRatePremium_,
+        uint256 optimalStableLoanRatio_
+    ) internal {
+        baseRatePerBlock = baseRatePerYear_ / blocksPerYear;
```

Comment:
do we need some validation for `baseRatePerYear_` if it is smaller than `blockPerYear` wouldn't that lead into loss of precision? I see this function is only called in constructor and admin function, but maybe some validation would be good?

Reply:
Added require checks for that please check in this [commit](https://github.com/VenusProtocol/venus-protocol/pull/210/commits/46f0cf168db69632d3f6a108f589965f5544c579)